### ISSUE TITLE
Migrate subtitles to VTT format

### ISSUE THREAD

**Issue by radinamatic** [learningequality/ka-lite#4166]
Migrate subtitles to VTT format

*milestone to be decided*
See #4072 and #4151 for context.

VideoJS version 4.11 that we currently have running on develop is the [last one that supports the](https://github.com/videojs/video.js/issues/1956#issuecomment-85617607) **SRT** format for subtitles. Given that the upgrade to `videojs` version 5 (once it is released) is recommended because of the a11y improvements, we'll have to implement the migration to **VTT** format prior to that.

**Resources**:
http://www.webvtt.org/
https://github.com/captioning/captioning/wiki#convert-a-subrip-file-srt-into-a-webvtt-file-vtt

**Comment by aronasorman** [comment]
Interesting. Since this has implications on our i18n pipeline, I'd prefer to do this later rather than sooner. Maybe 0.16?

**Comment by radinamatic** [comment]
Sounds OK to me, as we are in no hurry to implement it now.

(one more item for my post-GSoC list ;)

**Comment by radinamatic** [comment]
@aron This may be even easier than we thought, meaning not code-related. Amara (where our subs/transcripts are translated) allows to download subtitles in .vtt format. Not sure about the setup KA/FLE has over there, but I just checked on my Amara TED account and it lets me download my translations as .vtt files.

So if that's possible, we may even make it for 0.15? Implementing the transcript plugin will also depend on it as it's already ported to 4.12 videojs player, and we can't use it as-is.

**Comment by radinamatic** [comment]
Btw, [Video.JS **version 5.0** is released](http://blog.videojs.com/), so we should start considering how laborious is it going to be to re-download all our subs in VTT format from Amara... ;)

I can do that one by one from my dashboard there, but in our case (we are talking thousands), there must be a faster way!

![amara](https://cloud.githubusercontent.com/assets/1457929/10319759/40f1c47c-6c70-11e5-82be-8c7fd78a98af.png)
 

@aron @jamalex @rtibbles

**Comment by rtibbles** [comment]
@radinamatic Can you open an issue for migrating to VTT (if you didn't make one already)? @aron and I will be overhauling language packs somewhat this coming dev cycle, so this could fit into that.

**Comment by rtibbles** [comment]
Oh, I should really read the titles of things before I post...

**Comment by rtibbles** [comment]
............................................________........................ 
....................................,.-‘”...................``~.,.................. 
.............................,.-”...................................“-.,............ 
.........................,/...............................................”:,........ 
.....................,?......................................................\,..... 
.................../...........................................................,}.... 
................./......................................................,:`^`..}.... 
.............../...................................................,:”........./..... 
..............?.....__.........................................:`.........../..... 
............./__.(.....“~-,_..............................,:`........../........ 
.........../(_....”~,_........“~,_....................,:`........_/........... 
..........{.._$;_......”=,_.......“-,_.......,.-~-,},.~”;/....}........... 
...........((.....*~_.......”=-._......“;,,./`..../”............../............ 
...,,,___.\`~,......“~.,....................`.....}............../............. 
............(....`=-,,.......`........................(......;_,,-”............... 
............/.`~,......`-...............................\....../\................... 
.............\`~.*-,.....................................|,./.....\,__........... 
,,_..........}.>-._\...................................|..............`=~-,.... 
.....`=~-,_\_......`\,.................................\........................ 
...................`=~-,,.\,...............................\....................... 
................................`:,,...........................`\..............__.. 
.....................................`=-,...................,%`>--==``....... 
........................................_\..........._,-%.......`\............... 
...................................,<`.._|_,-&``................`\..............

**Comment by rtibbles** [comment]
![image](https://cloud.githubusercontent.com/assets/1680573/10319899/6e3c3b10-6c25-11e5-980b-9b0d1cbc4972.png)

**Comment by radinamatic** [comment]
Have a cookie, Captain... ;)

![Image](http://justjennrecipes.com/wp-content/uploads/2013/03/picard-facepalm-cookies.jpg)

**Comment by aronasorman** [comment]
Gio and I are looking into migrating to VTT this release.

**Comment by radinamatic** [comment]
Are you pulling all the subs as VTT and just changing the file type in `video` tag, or there is other magic involved?

**Comment by aronasorman** [comment]
@radinamatic I believe the former is the plan, as I'm hoping that videojs just magically interprets vtt :)

**Comment by radinamatic** [comment]
Nah... If you do that, magic is entirely yours @aronasorman, and VideoJS just makes use of it... :stuck_out_tongue_winking_eye: 

Ping me in on the PR so I can enjoy the magic too!

**Comment by benjaoming** [comment]
Haven't checked if you guys have been working on this, but I implemented a VTT convertion in the ka-lite-zim app:

https://github.com/benjaoming/ka-lite-zim/blob/436bdccd66f44aa9e521363b50d7560aa44f645d/kalite_zim/management/commands/export2zim.py#L268

The thing that made it work is [submarine](https://pypi.python.org/pypi/submarine/).

Furthermore, I met issues with having video.js automatically play with subtitles enabled. The fix is this script:

```javascript
    $(document).ready(function() {
      videojs("levideo").ready(function(){
        // Player (this) is initialized and ready.
        $(".vjs-subtitles-button .vjs-menu-item").eq(1).trigger('click');
      });
    });
```

**Comment by aronasorman** [comment]
The english content packs now come with `vtt`. Next step is to check if videojs reads the subtitles.

**Comment by MCGallaspy** [comment]
> The english content packs now come with vtt. Next step is to check if videojs reads the subtitles.

How can I confirm this?

**Comment by rtibbles** [comment]
Check that subtitles are available on an English video? If not, probably an
issue as most videos are subtitled.

**Comment by aronasorman** [comment]
Verified that subtitles are present in the latest `en.zip` in pantry. They should get extracted with the `.vtt` extension.

**Comment by aronasorman** [comment]
Are subtitles confirmed to be working on `en` now?

**Comment by MCGallaspy** [comment]
My mistake, I read your last comment to mean that this was fixed!

**Comment by MCGallaspy** [comment]
Subtitles are included but we're not serving them: https://github.com/learningequality/ka-lite/blob/0.16.x/kalite/topic_tools/annotate.py#L112

**Comment by MCGallaspy** [comment]
Potentially try upgrading video.js as well: https://github.com/learningequality/ka-lite/blob/0.16.x/package.json#L75

**Comment by rtibbles** [comment]
Fixed in #4943
